### PR TITLE
[MRG] Better initialization

### PR DESCRIPTION
**Pull request by AlexanderFabisch**
Improved initialization with
* k-means++
* initial covariance scaling

Figure_1

### CODE REVIEW

File: gmr/gmm.py [hunk 1]
```diff
@@ -46,6 +51,35 @@ def kmeansplusplus_initialization(X, n_components, random_state=None):
...
+        Initial covariances
+    """
+    n_features = X.shape[1]
+    average_distances = np.empty(n_features)
+    for i in range(n_features):
+        average_distances[i] = np.mean(
+            pdist(X[:, i, np.newaxis], metric="euclidean"))
+    initial_covariances = np.empty((n_components, n_features, n_features))
+    initial_covariances[:] = np.eye(n_features) * (average_distances / n_components) ** 2
+    return initial_covariances
```
[the remark this answers is not in the document]
Comment (AlexanderFabisch) [reply gmr/gmm.py:84]: Yes, try it yourself. The example script that generates the plots that you see in the description. Actually, in this specific example, both initialization methods k-means++ and covariance scaling on their own are sufficient to fix the issue. Both combined should be even more reliable.

File: gmr/gmm.py [hunk 2]
```diff
@@ -1,8 +1,89 @@
...
+
+    random_state = check_random_state(random_state)
+
+    all_indices = np.arange(len(X))
+    selected_centers = [random_state.choice(all_indices, size=1).tolist()[0]]
+    while len(selected_centers) < n_components:
+        centers = np.atleast_2d(X[np.array(selected_centers, dtype=int)])
+        squared_distances = cdist(X, centers, metric="sqeuclidean")
+        selection_probability = squared_distances.max(axis=1)
+        selection_probability[np.array(selected_centers, dtype=int)] = 0.0
```
[the remark this answers is not in the document]
Comment (AlexanderFabisch) [reply gmr/gmm.py:47]: > I don't understand why this is necessary

If you mean this particular line: to avoid sampling the same example more than once.

> In a more general matter, why not using the Kmeans fro sklearn for initialization?

That would add another dependency. I'm thinking about mentioning that gmr is fully compatible to sklearn though. I will probably add a new example script soon and mention it in the readme.

You can even use sklearn's Dirichlet process GMM to "initialize" priors, means, and covariances.

[k-means++ initialization](https://en.wikipedia.org/wiki/K-means%2B%2B#Improved_initialization_algorithm) is a well-known initialization method for k-means and GMMs. It requires less computational power than k-means, which might be an advantage for large datasets. Nevertheless, I don't know whether k-means initialization + less EM iterations or k-means++ initialization + more EM iterations is more expensive... I'll let the user decide.

File: examples/plot_regression.py [hunk 3]
```diff
@@ -20,7 +20,7 @@
 n_samples = 10
 X = np.ndarray((n_samples, 2))
 X[:, 0] = np.linspace(0, 2 * np.pi, n_samples)
-X[:, 1] = 1 - 3 * X[:, 0] + random_state.randn(n_samples)
+X[:, 1] = 100 * (1 - 3 * X[:, 0] + random_state.randn(n_samples))
```
Comment (AlexanderFabisch) [inline examples/plot_regression.py:23]: Revert this.

File: gmr/gmm.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Labulitiolle) [reply gmr/gmm.py:47]: For the first point: My bad, forgot about the loop.
For the second point: From what I remember EM has more parameters (N*(N+1)/2) to optimize than Kmeans thus the complexity will be higher I guess?

Do you think using hard K-means would improve the model?

[the remark this answers is not in the document]
Comment (AlexanderFabisch) [reply gmr/gmm.py:47]: Standard k-means doesn't have a covariance. Using k-means for initialization is certainly a good way to initialize the means of a GMM. sklearn does it [by default](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/mixture/_base.py#L145).

[the remark this answers is not in the document]
Comment (Labulitiolle) [reply gmr/gmm.py:47]: Yes my bad, I thought you used Soft K-means for your above picture.
